### PR TITLE
feat(browser): support allowed_private_hosts for browser tools

### PR DESCRIPTION
**Pull request by NiuBlibing**
## Summary

- **Base branch:** `master`
- **What changed and why:**
  - Add `[browser].allowed_private_hosts` so the `browser` and `browser_open` tools can reach operator-listed private/LAN hosts — previously *any* private host was unconditionally blocked, with no escape hatch (only `http_request`/`web_fetch` had one). This unblocks legitimate internal-service automation.
  - Mirror the existing `http_request` design: exact/subdomain matches, and `["*"]` permits **all** private/local hosts.
  - Listed private hosts bypass the SSRF block *and* the `allowed_domains` allowlist; the wildcard is scoped to private hosts only and never widens public reach.
- **Scope boundary:** Does **not** loosen URL schemes — `browser_open` stays HTTPS-only and `file://` stays blocked everywhere. Does **not** touch `http_request`/`web_fetch`/`text_browser`/`browser_delegate` or the shared `domain_guard` SSRF logic. No DNS-rebinding posture change.
- **Blast radius:** `browser`/`browser_open` tool registration in `zeroclaw-runtime`; the `[browser]` config surface. Deny-by-default means zero behavior change unless an operator opts in.
- **Linked issue(s):** None.
- **Labels:** `risk: high` (touches `crates/zeroclaw-tools/src/**` SSRF access-control boundary per AGENTS.md), `size: M` — maintainers own final risk/size.

## Validation Evidence (required)

```bash
cargo fmt --all -- --check
cargo clippy --workspace --all-targets -- -D warnings
cargo test -p zeroclaw-tools -p zeroclaw-config -p zeroclaw-runtime --lib --tests
```

- **Commands run and tail output:**
  - `cargo fmt --all -- --check` → clean, exit 0.
  - `cargo clippy --workspace --all-targets -- -D warnings` → `Finished \`dev\` profile ... in 13m 37s`, exit 0, zero warnings across all crates.
  - Tests (lib + integration):
    - `zeroclaw-tools` lib: `test result: ok. 1353 passed; 0 failed`
    - `zeroclaw-config` lib: `ok. 942 passed; 0 failed`; `tests/migration.rs`: `ok. 90 passed`; `tests/comment_writer.rs`: `ok. 8 passed`
    - `zeroclaw-runtime` lib: `2342 passed; 1 failed` — the one failure is `cron::store::tests::remove_job_emits_structured_cron_delete_event`, **unrelated to this diff** (cron module). It **passes in isolation** (`ok. 1 passed`); it shares global log-capture state (`__private_test_writer_lock` / `subscribe_or_install`) and races other tests under parallel run. Pre-existing flake.
- **Beyond CI — what did you manually verify?** Hand-traced every `validate_url` branch in both tools for {public,private}×{listed,unlisted}×{http,https}. Added adversarial tests: wildcard private allowlist does **not** widen the public `allowed_domains` allowlist; `file://` stays blocked even with `["*"]`; empty list still denies all private hosts; private host bypasses `allowed_domains` only when listed. **Not verified:** live agent-browser/system-browser execution against a real internal host — validation is at the URL-gate layer; no live network was exercised.
- **If any command was intentionally skipped, why:** `cargo test --doc` for `zeroclaw-config` skipped — pre-existing rustdoc env failure (`error: Option 'default-theme' given more than once`; rustdoc invoked with `--default-theme=ayu` twice), reproduced on a clean tree via `git stash`, independent of this change.

## Security & Privacy Impact (required)

- New permissions, capabilities, or file system access scope? **Yes** — adds an opt-in capability for `browser`/`browser_open` to reach operator-listed private/LAN hosts that were previously always blocked. Off by default (empty list = deny).
- New external network calls? **No** — no new call sites; only which hosts pass validation changes.
- Secrets / tokens / credentials handling changed? **No**.
- PII, real identities, or personal data in diff, tests, fixtures, or docs? **No** — fixtures/tests use only RFC1918 addresses, `localhost`, and `example.com`.
- **Risk & mitigation:** This is an SSRF-control relaxation, so it is strictly **opt-in and deny-by-default**. The wildcard is **scoped to private hosts only** — a `["*"]` in `allowed_private_hosts` never grants access to public hosts (those still flow through `allowed_domains`). `file://` remains blocked; `browser_open` remains HTTPS-only. No change to DNS-rebinding posture or the shared `domain_guard` matcher.

## Compatibility (required)

- Backward compatible? **Yes** — new field defaults to empty; existing configs behave exactly as before.
- Config / env / CLI surface changed? **Yes** — adds `[browser].allowed_private_hosts` (`Vec<String>`, default `[]`).
- **Upgrade steps:** None required. To opt in, set `[browser].allowed_private_hosts = ["10.0.0.5", "internal.local"]` (or `["*"]` for all private hosts).

## Rollback (required for `risk: medium` and `risk: high`)

- **Fast rollback command/path:** `git revert 3fe8e6b5b2`.
- **Feature flags or config toggles:** `[browser].allowed_private_hosts` — set to `[]` (the default) to fully disable; no rebuild needed.
- **Observable failure symptoms:** `browser`/`browser_open` reaching unexpected internal hosts; in logs, an unexpected drop in `Blocked local/private host` rejections or tool-open events targeting RFC1918 / link-local / metadata addresses.

### CODE REVIEW

File: crates/zeroclaw-config/src/schema.rs [hunk 1]
```diff
@@ -6618,6 +6618,13 @@ pub struct BrowserConfig {
     #[serde(default)]
     #[nested]
     pub computer_use: BrowserComputerUseConfig,
+    /// Private/internal hosts allowed to bypass SSRF protection.
+    /// Exact and subdomain matches are supported; `["*"]` permits **all** private/local
+    /// hosts (RFC 1918, loopback, link-local, `.local`). Default: empty (deny).
+    /// Listed hosts also bypass `allowed_domains` and (for `browser_open`) the
```
Comment (singlerider) [inline crates/zeroclaw-config/src/schema.rs]: This doc promises that listed private hosts bypass `browser_open`'s HTTPS-only restriction, but `browser_open::validate_url` rejects any non-`https://` URL before the `allowed_private_hosts` logic runs. Either implement the bypass (allow `http://` for listed private hosts + add an `http://` test) or strike the "(for browser_open) the HTTPS-only restriction" clause. See review body.

**Review by singlerider — CHANGES_REQUESTED** [review]
The design is clean: it mirrors the established `http_request` precedent exactly (`domain_guard::normalize_allowed_domains` + `host_matches_allowlist`), deny-by-default, wildcard scoped to private hosts only, and the adversarial test matrix is genuinely good. One blocker on a doc-vs-code contradiction, then this is landable.

### 🔴 Blocking — schema doc claims a `browser_open` HTTPS-only bypass that the code does not implement

The new field doc in `crates/zeroclaw-config/src/schema.rs` states:

> Listed hosts also bypass `allowed_domains` and (for `browser_open`) the HTTPS-only restriction — internal services frequently lack a public TLS cert.

But in `browser_open.rs::validate_url`, the scheme guard runs first and unconditionally:

```rust
if !url.starts_with("https://") {
    anyhow::bail!("Only https:// URLs are allowed");
}
// ... allowed_private_hosts logic only runs AFTER this point
```

So `http://internal.local` against a listed private host is still rejected. The capability the doc advertises (reach internal services that lack a public TLS cert) does not exist — and an operator reading the schema will configure `allowed_private_hosts` expecting plain-`http` internal services to work, then hit a hard `Only https:// URLs are allowed` with no indication why.

Two acceptable resolutions, your call:

1. **Implement the bypass** the doc promises: allow `http://` for hosts that match `allowed_private_hosts` in `browser_open` (move the private-host classification ahead of the scheme guard, or special-case it), and add a test proving `http://10.0.0.1` passes for a listed host. This is the more useful behavior given the stated motivation.
2. **Fix the doc** to match the code: drop the "(for `browser_open`) the HTTPS-only restriction" clause, since `browser_open` stays HTTPS-only regardless of the allowlist.

Whichever you pick, the `browser_open` private-host tests should include an `http://` case so the scheme/allowlist interaction is actually pinned — right now every `browser_open` private-host test uses `https://`, so this gap is untested in either direction.

### 🟢 Praise — test matrix and precedent fidelity

The adversarial coverage is the right set: wildcard-does-not-widen-public-allowlist, `file://`-stays-blocked-under-`["*"]`, empty-list-still-denies, listed-vs-unlisted. Reusing `domain_guard`'s shared matcher instead of reimplementing host matching keeps `browser` aligned with `http_request`/`web_fetch` and means the SSRF classifier (`is_private_or_local_host`, including link-local and IPv4-mapped IPv6) stays single-sourced. The `new` → `new_with_private_hosts` shim keeps the old constructor call sites compiling without churn.

### 🟢 Note (non-blocking) — `["*"]` exposes the cloud metadata endpoint, by design and consistent with precedent

`is_private_or_local_host` classifies link-local (`169.254.0.0/16`) as private, so a `["*"]` wildcard makes `169.254.169.254` (cloud instance metadata) reachable by the browser tools. This is the *same* exposure `http_request`'s `["*"]` already grants, so it's not a new posture — and it's opt-in and off by default. Not gating. Worth one sentence in the field doc warning that `["*"]` includes link-local/metadata, so an operator reaching for `["*"]` for convenience understands the metadata-SSRF surface they're accepting.

Local validation at head: `cargo test -p zeroclaw-tools --lib` private-allowlist set — 12 passed, 0 failed.

**Comment by NiuBlibing** [comment]
@singlerider thanks for the careful read — you're right that the doc advertised a capability the code never implemented. I went with your **option 2 (fix the doc to match the code)** rather than option 1, to keep this PR scoped to its title (the `allowed_private_hosts` SSRF allowlist) and leave `browser_open`'s scheme policy to land separately in #8136. Pushed in c8a7cd487e.

Concretely:

- **Struck the false bypass claim.** The field doc no longer says listed hosts bypass `browser_open`'s HTTPS-only restriction. It now states the actual split behavior: the `browser` tool accepts `http://` for listed hosts (its scheme guard already allows `http`/`https`, and the private-host bypass returns before the allowlist check), while **`browser_open` stays HTTPS-only regardless of this list** — its scheme guard rejects non-`https` before the `allowed_private_hosts` logic ever runs.
- **Added the link-local/metadata warning** you flagged in the 🟢 note: the doc now warns that `["*"]` reaches link-local addresses including the cloud metadata endpoint (`169.254.169.254`), so an operator reaching for `["*"]` understands the surface.
- **Pinned the scheme/allowlist interaction** in `browser_open` tests: new `listed_private_host_still_rejects_http_scheme` asserts `http://10.0.0.1` (a listed host) is rejected by the scheme guard, not the allowlist. This closes the untested-in-either-direction gap you called out.

Validation on this branch: `cargo fmt --all -- --check` clean; `cargo clippy -p zeroclaw-tools --all-targets -- -D warnings` clean; `cargo test -p zeroclaw-tools --lib browser_open::` → 21 passed, 0 failed.

File: crates/zeroclaw-config/src/schema.rs [hunk 2]
```diff
@@ -6822,6 +6822,16 @@ pub struct BrowserConfig {
     #[serde(default)]
     #[nested]
     pub computer_use: BrowserComputerUseConfig,
+    /// Private/internal hosts allowed to bypass SSRF protection.
+    /// Exact and subdomain matches are supported; `["*"]` permits **all** private/local
+    /// hosts (RFC 1918, loopback, link-local, `.local`). Default: empty (deny).
+    /// Listed hosts also bypass `allowed_domains`. The `browser` tool accepts `http://`
+    /// for listed hosts (internal services frequently lack a public TLS cert); `browser_open`
+    /// stays HTTPS-only regardless of this list.
```
Comment (singlerider) [inline crates/zeroclaw-config/src/schema.rs]: 🟢 This is the exact resolution. The doc now matches the code: `browser` accepts http for listed private hosts, `browser_open` stays HTTPS-only. No more phantom capability for an operator to trip over.

File: crates/zeroclaw-tools/src/browser_open.rs [hunk 3]
```diff
@@ -441,4 +480,72 @@ mod tests {
...
+    }
+
+    #[test]
+    fn specific_private_host_alone_satisfies_allowlist_requirement() {
+        let tool = test_tool_with_private(vec![], vec!["192.168.1.5"]);
+        assert!(tool.validate_url("https://192.168.1.5").is_ok());
+    }
+
+    #[test]
+    fn listed_private_host_still_rejects_http_scheme() {
```
Comment (singlerider) [inline crates/zeroclaw-tools/src/browser_open.rs]: 🟢 Pins the HTTPS-only invariant for browser_open against a listed private host (`http://10.0.0.1` -> rejected). Combined with browser.rs accepting the same http URL for a listed host, the per-tool scheme split is now tested in both directions — closes the gap I flagged.

**Review by singlerider — APPROVED** [review]
Re-reviewed at `c8a7cd487e`. Built `zeroclaw-tools` and ran the full lib suite: `1379 passed, 0 failed`. (The raw diff against my prior CR SHA is a large rebase onto a moved master; I diffed against the real merge-base `90a791e582` to isolate this PR's actual changes.)

### RESOLVED ✅ Doc-vs-code contradiction fixed, and pinned both directions

My blocker was the schema doc claiming `browser_open` bypasses the HTTPS-only restriction for listed private hosts, which the code did not implement. You took resolution 2 (fix the doc) and did it precisely. The field doc now reads:

> The `browser` tool accepts `http://` for listed hosts (internal services frequently lack a public TLS cert); `browser_open` stays HTTPS-only regardless of this list.

That matches the code exactly. `browser_open::validate_url` keeps the unconditional `https://` guard ahead of the private-host logic, while `browser::validate_url` accepts both schemes and gates a private host on `allowed_private_hosts`.

Both interactions are now pinned with tests, which is exactly the gap I asked to close:

- `browser_open::tests::listed_private_host_still_rejects_http_scheme` — `http://10.0.0.1` to a listed host is still rejected. Pins "browser_open stays HTTPS-only regardless."
- `browser::tests::allowed_private_hosts_entry_permits_listed_host` plus the `http://192.168.1.5` / `http://10.0.0.1` ok-cases and the `http://10.0.0.2` unlisted-reject — pins that the `browser` tool accepts `http://` for listed private hosts only.

### RESOLVED ✅ Metadata/link-local warning added

My non-blocking note asked for one sentence warning that `["*"]` reaches link-local including the cloud metadata endpoint. The doc now carries it verbatim in intent:

> Warning: `["*"]` also reaches link-local addresses, including the cloud metadata endpoint (`169.254.169.254`) — list specific hosts unless you accept that exposure.

The design fidelity I praised last round still holds: shared `domain_guard` matcher (SSRF classifier single-sourced with `http_request`/`web_fetch`), deny-by-default, `file://` stays blocked under `["*"]`, `new` -> `new_with_private_hosts` shim. The adversarial matrix is intact and now complete on the scheme axis.

Converting my CR to approve.

**Review by Audacity88 — CHANGES_REQUESTED** [review]
Reviewed current head c8a7cd487e against live PR state, the full PR body, comments, inline threads, @singlerider's prior reviews, the diff against merge base 90a791e582, and the relevant browser/config/runtime source. I did not run local Cargo validation or a live browser-backend smoke in this review-only pass.

### ✅ Resolved — the `browser_open` scheme contract now matches the code

I agree with @singlerider's re-review that the original doc/code mismatch is resolved. The schema no longer promises that `allowed_private_hosts` lets `browser_open` bypass its HTTPS-only guard, and `browser_open::tests::listed_private_host_still_rejects_http_scheme` pins that `http://10.0.0.1` still fails even when `10.0.0.1` is listed. The `browser` tests also pin the other side of the split: the browser automation tool can use `http://` for listed private hosts while unlisted private hosts still fail.

### 🔴 Blocking — `browser` still lets userinfo hide the private host from the SSRF gate

`BrowserTool::validate_url` still feeds the raw URL to the browser backend after checking a host string produced by the hand-rolled `extract_host` helper. That helper strips the scheme, takes the authority before `/`, and then takes everything before `:` for non-IPv6 hosts. It never rejects `@` userinfo.

That creates a parser mismatch. Under normal URL semantics, `http://example.com@127.0.0.1/` has userinfo `example.com` and host `127.0.0.1`. The browser helper instead classifies the string `example.com@127.0.0.1`. That string is not the loopback host the browser will navigate to, and when `browser.allowed_domains` permits `["*"]`, the allowlist check accepts it. The browser backend then receives the original URL, whose actual destination is loopback. That bypasses the SSRF guard this PR is extending.

This parser hole existed before this branch, but this PR is editing the same browser SSRF gate and adding adversarial coverage for the new private-host bypass. Before this lands, `browser` should either parse with the same URL semantics the browser will use or reject userinfo explicitly, as `browser_open` already does with `authority.contains('@')`. Please also add a regression test that a userinfo URL targeting a private host is rejected, including the default `["*"]` public allowlist case; adding the same check with `allowed_private_hosts = ["*"]` would make the new bypass boundary explicit too.

**Comment by NiuBlibing** [comment]
@Audacity88 thanks — you're right that `extract_host` was a hand-rolled prefix-strip that never rejected `@`, so `http://example.com@127.0.0.1/` was being classified by the `example.com@…` string (which satisfies the default `["*"]` allowlist) while the browser navigated to loopback. Pushed in 8fc343b5b8.

Concretely:

- **Reject userinfo before host extraction.** `BrowserTool::validate_url` now mirrors `browser_open::validate_url`'s authority check — strip the scheme, take the authority before `/`/`?`/`#`, reject if it contains `@`. The check sits right after the scheme guard and before `extract_host`, so the parser-mismatch is closed at the same gate the allowlist uses.
- **Pinned both allowlist surfaces with regression tests** so neither can be used to smuggle a private destination:
  - `userinfo_url_targeting_private_host_rejected_under_wildcard_public_allowlist` — `http://example.com@127.0.0.1/` rejected under the **shipped default** `allowed_domains = ["*"]`, no private allowlist.
  - `userinfo_url_targeting_private_host_rejected_under_wildcard_private_allowlist` — same URL rejected under `allowed_private_hosts = ["*"]`, making the new bypass boundary explicit (userinfo is a parser-mismatch defense, not something an operator can opt around by opening the private bypass wide).
  - `userinfo_url_with_password_rejected` — `user:pass@host` form covered as well.

I chose the explicit-reject path over reaching for a full URL parser because (a) it matches `browser_open`'s existing posture verbatim, (b) it keeps the `extract_host` helper single-sourced with the SSRF guard, and (c) userinfo on a browser URL has no legitimate use here — `browser_open` rejects it unconditionally and `browser` should too.

Validation on this branch: `cargo fmt --all -- --check` clean; `cargo clippy -p zeroclaw-tools --all-targets -- -D warnings` clean; `cargo test -p zeroclaw-tools --lib browser::` — **54 passed, 0 failed**, including the 3 new userinfo regression tests.

**Review by Audacity88 — CHANGES_REQUESTED** [review]
I reviewed current head 8fc343b5b8a1ff308b673a28789a15cf697402f1 against the live PR state, current checks, @singlerider's approval, my prior request-changes review, the author's follow-up comments, and the current `browser`, `browser_open`, config schema, runtime tool-registration, and `domain_guard` source. I did not run local Cargo or a live browser smoke in this review-only pass; all visible GitHub checks are green.

### ✅ Resolved — the prior userinfo bypass is fixed in the right place

@NiuBlibing, thanks for the careful follow-up. The new `BrowserTool::validate_url` userinfo check addresses the exact parser mismatch from my prior review: `http://example.com@127.0.0.1/` is now rejected before `extract_host`, and the tests cover both the shipped public wildcard allowlist and the new private wildcard allowlist. I also agree with @singlerider that the earlier `browser_open` scheme/documentation mismatch is resolved: the config docs now say `browser_open` stays HTTPS-only, and the test pins that listed private hosts do not bypass that scheme guard.

### 🔴 Blocking — `browser` host extraction still needs the same authority boundary

There is one more parser mismatch in the same `browser` SSRF gate. The new userinfo precheck correctly treats `/`, `?`, and `#` as authority boundaries before it checks for `@`. But the existing `extract_host` helper used for the actual private-host and allowlist decision is still the narrower helper described in my prior review and confirmed in the author's follow-up: it strips the scheme, takes the authority before `/`, and then takes everything before `:` for non-IPv6 hosts.

That means a browser URL with no slash before the query or fragment can still hide the real private host from the guard. For example, `http://127.0.0.1?x` can be classified by the helper as host `127.0.0.1?x`, not `127.0.0.1`. That string is not classified as private/local, so with the default public wildcard allowlist it can pass `allowed_domains`, while the browser backend receives the original URL and navigates to loopback. The same shape applies to `#` fragments.

This is the same class of issue as the userinfo bug: the security check and the browser's URL parser do not agree on the authority boundary. Since this PR is changing the browser SSRF boundary and adding adversarial coverage for it, please make `browser::extract_host` use the same `/`, `?`, and `#` authority boundary as the new precheck, or otherwise parse the host with URL semantics that match the browser target. Please also add regressions showing that `http://127.0.0.1?x` and `http://127.0.0.1#x` are rejected under the default `allowed_domains = ["*"]` posture when `allowed_private_hosts` is empty.

Once that remaining separator mismatch is closed, this should be in good shape from my side.

**Comment by NiuBlibing** [comment]
@Audacity88 thanks — you're right that fixing the userinfo case while keeping the hand-rolled `extract_host` helper just left the door open for whatever the next authority boundary it missed (`?`, `#`, …). Pushed in 25beebaec3.

Rather than add a third separator to the hand-rolled splitter, I switched `BrowserTool::validate_url` to parse with **`reqwest::Url`** (re-exported `url` crate — the Rust de-facto standard URL parser, already used in this crate by `http_request.rs`, `domain_guard.rs`, and elsewhere in `browser.rs` itself for the WebDriver endpoint). So the security check now uses the same parser the browser backend ultimately resolves against — the parser-mismatch class is closed at the root, not at the next-discovered separator.

Concretely:

- **Replaced `extract_host` with `reqwest::Url::parse`.** Authority/userinfo/host are now extracted by the URL crate. Removed the hand-rolled helper and its unit tests — the IPv6 / port / case-normalization behavior is delegated to `url` and end-to-end covered by `validate_url_blocks_ipv6_ssrf` plus the existing allowlist tests.
- **Userinfo check now uses `parsed.username()` / `parsed.password()`** instead of a manual `contains('@')`, matching `http_request.rs`'s posture verbatim. The userinfo regression tests from the prior round still pass unchanged.
- **Re-add IPv6 brackets after `Url::host_str`** so the host string fed to `is_private_or_local_host` and `host_matches_allowlist` keeps the `[::1]` shape used elsewhere in this crate. Avoids depending on `url::Host` (only `url::Url` is re-exported via `reqwest`).
- **Pinned both flagged shapes end-to-end through `validate_url`** under the shipped `allowed_domains = ["*"]` posture with no private allowlist:
  - `query_only_url_targeting_private_host_rejected_under_wildcard_public_allowlist` — `http://127.0.0.1?x` blocked as private host.
  - `fragment_only_url_targeting_private_host_rejected_under_wildcard_public_allowlist` — `http://127.0.0.1#x` blocked as private host.

Validation on this branch:

```
cargo fmt --all -- --check                                  # clean
cargo clippy -p zeroclaw-tools --all-targets -- -D warnings # clean
cargo test -p zeroclaw-tools --lib browser::                # 54 passed, 0 failed
cargo test -p zeroclaw-tools --lib                          # 1382 passed, 0 failed (was 1379)
```

**Comment by NiuBlibing** [comment]
Heads-up on the failing `Test` job — that one wasn't from the `reqwest::Url` switch, it's a master-drift issue I fixed in 3c8493742d:

- Master has since relaxed `browser_open`'s scheme guard to accept **both** `http://` and `https://` (see `!(starts_with("https://") || starts_with("http://"))` in `browser_open::validate_url` on master).
- After I rebased this PR onto current master, the test `listed_private_host_still_rejects_http_scheme` (added back in c8a7cd487e to pin the old HTTPS-only posture for `browser_open`) started failing in CI with `called Result::unwrap_err() on an Ok value: "http://10.0.0.1"` — it asserts a contract that no longer exists.

Two surfaces needed to follow that change:

1. **Flipped the test to pin the new behavior.** Renamed to `listed_private_host_permits_http_scheme` and now asserts that `http://10.0.0.1` is accepted for a **listed** private host (since the scheme guard now allows it). The unlisted-host SSRF guard is still covered separately by `allowed_private_hosts_does_not_permit_unlisted_host`.
2. **Updated the `[browser].allowed_private_hosts` schema doc** so it matches master: was "`browser_open` stays HTTPS-only regardless of this list," now "Both the `browser` tool and `browser_open` accept `http://` for listed hosts." The link-local/metadata warning is kept verbatim.

Validation on the rebased branch:

```
cargo fmt --all -- --check                                  # clean
cargo test -p zeroclaw-tools --lib browser_open::           # 24 passed, 0 failed
cargo test -p zeroclaw-tools -p zeroclaw-config --lib      # 1451 passed, 0 failed
```

CI is green.

**Review by Audacity88 — APPROVED** [review]
Reviewed current head d7c09211868476ca76c1d95c46574932963987f4 against the live PR state, current green checks, the full prior review thread, the author follow-up comments, and the current `browser`, `browser_open`, config schema, runtime tool-registration, and shared domain-guard source. I did not run local Cargo or a live browser/system-browser smoke in this review-only pass.

### ✅ Resolved — `browser` now uses URL-parser authority semantics before the SSRF decision

@NiuBlibing, thanks for closing the remaining parser mismatch rather than patching only the examples. The current `BrowserTool::validate_url` now parses with `reqwest::Url`, rejects userinfo before host classification, and takes the effective `host_str()` from the parsed URL before applying the private-host and public-domain checks. That resolves both blockers I had raised: `http://example.com@127.0.0.1/` is rejected as userinfo, and query/fragment-only authority boundaries like `http://127.0.0.1?x` / `http://127.0.0.1#x` are classified as loopback instead of being allowed through the default public wildcard.

The regression coverage pins the important postures too: userinfo is rejected under both public-wildcard and private-wildcard allowlists, and the query/fragment loopback shapes are rejected under the shipped `allowed_domains = ["*"]` posture when `allowed_private_hosts` is empty.

### ✅ Resolved — `browser_open` now matches the current master scheme policy

The earlier review thread briefly had an HTTPS-only `browser_open` contract, but master has since relaxed `browser_open` to accept both `http://` and `https://`. This head now matches that current policy: the schema field doc says both `browser` and `browser_open` accept `http://` for listed private hosts, and the `browser_open` test pins `http://10.0.0.1` as allowed when that private host is listed while unlisted private hosts remain blocked.

### 🟡 Warning — exact IPv6 private-host entries look fail-closed

One edge case still looks worth following up on, but I am not blocking on it because it fails closed. `BrowserTool::validate_url` re-adds brackets before feeding IPv6 literals into `host_matches_allowlist`, while `normalize_allowed_domains` stores IPv6 allowlist entries without brackets. That means `allowed_private_hosts = ["[fd00::1]"]` or `["fd00::1"]` appears unlikely to match `http://[fd00::1]/` as a specific entry. This is a specific-IPv6 support gap, not a reopened SSRF path.

If we intend specific IPv6 private hosts to be supported, the follow-up should add the missing specific-IPv6 allowlist test and normalize the match shape. This does not reopen the SSRF bug class from my prior reviews.

### 🟡 Warning — model-facing tool descriptions still describe the old public-only posture

The runtime checks are enforcing the new allowlist correctly, but the tool descriptions are slightly stale: `browser_open::description()` still says "no local/private hosts", and `browser::description()` only mentions `browser.allowed_domains` for open actions. That can make the new operator-configured path less discoverable to models/users even though the validation path is correct. Those descriptions should be cleaned up before or shortly after merge, but I do not consider that a blocker for this security boundary.

### 🟢 What looks good — the new bypass remains scoped and opt-in

The core shape is sound now: default `allowed_private_hosts = []` preserves the existing private-host deny behavior, listed private hosts bypass `allowed_domains` only after being classified private/local, public hosts still flow through `browser.allowed_domains`, and `file://` remains blocked even under wildcard private-host allowlists. Runtime registration also passes the new config into both `browser` and `browser_open`, so the schema knob reaches the actual tool instances.

Approving to clear my stale request-changes review. This is not a merge-ready statement: the PR body still needs a maintainer-side cleanup before merge because it contains older validation and scheme-policy wording from before the final master rebase.
